Changes in module 'nf-core/pigz/uncompress'
--- modules/nf-core/pigz/uncompress/main.nf
+++ modules/nf-core/pigz/uncompress/main.nf
@@ -42,7 +42,7 @@
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
-        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\w*//' ))
+        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\\w*//' ))
     END_VERSIONS
     """
 }

************************************************************
